tests/jdom/tester.ts

```diff
@@ -0,0 +1,193 @@
+    })
+    return true
+}
+
+
+interface ServerDevice {
+    server: JDServiceServer,
+    roleName?: string,
+}
+
+
+// Creates a bus with the specified servers (optionally bound to the specified roles),
+// then runs the test function.
+export async function withBus(devices: ServerDevice[], 
+    test: (bus: JDBus, serviceMap: Map<JDServiceServer, JDService>) => Promise<void>) {
+    const bus = mkBus()
+
+    // For server devices: add the service provider on the bus and return the device
+    const serverDevices = devices.map(device => {
+        const busDevice = bus.addServiceProvider(new JDServiceProvider(
+            [(<ServerDevice>device).server] 
+        ))
+        return {
+            server: (<ServerDevice>device).server,
+            busDevice: busDevice,
+            roleName: (<ServerDevice>device).roleName,
+        }
+    })
+
+    bus.start()
+
+    // Wait for created devices to be announced, so services become available
+    const serverDeviceIds = serverDevices.map(elt => {
+        return elt.busDevice.deviceId
+    })
+    await new Promise(resolve => {
+        const devicesIdSet = new Set(serverDeviceIds)
+        const announcedIdSet = new Set()
+        const onHandler = (device: JDDevice) => {
+            if (devicesIdSet.has(device.deviceId)) {
+                announcedIdSet.add(device.deviceId)
+            }
+            if (setEquals(devicesIdSet, announcedIdSet)) {
+                bus.off(DEVICE_ANNOUNCE, onHandler)
+                resolve(undefined)
+            }
+        }
+        bus.on(DEVICE_ANNOUNCE, onHandler)
+    })
+
+    // Bind services to roles
+    // Start by geting server service -> roleName mappings
+    const serverServiceRoles = serverDevices.map(elt => {
+        const services = elt.busDevice.services({serviceClass: elt.server.serviceClass})
+        assert(services.length > 0, 
+            `created device ${elt.busDevice.friendlyName} has no service of ${elt.server.specification.name}`)
+        assert(services.length == 1, 
+            `created device ${elt.busDevice.friendlyName} has multiple service of ${elt.server.specification.name}`)
+        return {
+            service: services[0],
+            roleName: elt.roleName
+        }
+    })
+
+    const roleBindings = serverServiceRoles.filter(elt => {
+        return !!elt.roleName  // filter for where role name is available
+    }).map(elt => {
+        return {
+            role: elt.roleName, 
+            serviceClass: elt.service.specification.classIdentifier,
+            service: elt.service,
+        }
+    })
+
+    const roleManager = new RoleManager(bus)
+    roleManager.setRoles(roleBindings)
+
+    // Return created services as a map from the source server
+    // Trace devices are ignored here, since handles aren't (currently?) useful
+    const serviceMap = new Map(
+        serverDevices.map(elt => {
+            return [elt.server, elt.busDevice.services({serviceClass: elt.server.serviceClass})[0]]
```

Comment:
Use an object rather an array here. Easier to understand the data structures moving around.

Reply:
Can the Map constructor take objects for the map entries?

Reply:
Oh i see, the array pairs get unfolded in the map.